--!strict

type Instances = {
	Instance: Instance,
	Accoutrement: Accoutrement,
	Accessory: Accessory,
	Hat: Hat,
	AdPortal: AdPortal,
	AdService: AdService,
	AdvancedDragger: AdvancedDragger,
	AnalyticsService: AnalyticsService,
	Animation: Animation,
	AnimationClip: AnimationClip,
	CurveAnimation: CurveAnimation,
	KeyframeSequence: KeyframeSequence,
	AnimationClipProvider: AnimationClipProvider,
	AnimationController: AnimationController,
	AnimationFromVideoCreatorService: AnimationFromVideoCreatorService,
	AnimationFromVideoCreatorStudioService: AnimationFromVideoCreatorStudioService,
	AnimationRigData: AnimationRigData,
	AnimationStreamTrack: AnimationStreamTrack,
	AnimationTrack: AnimationTrack,
	Animator: Animator,
	AppUpdateService: AppUpdateService,
	AssetCounterService: AssetCounterService,
	AssetDeliveryProxy: AssetDeliveryProxy,
	AssetImportService: AssetImportService,
	AssetImportSession: AssetImportSession,
	AssetManagerService: AssetManagerService,
	AssetPatchSettings: AssetPatchSettings,
	AssetService: AssetService,
	Atmosphere: Atmosphere,
	Attachment: Attachment,
	Bone: Bone,
	AudioAnalyzer: AudioAnalyzer,
	AudioChorus: AudioChorus,
	AudioCompressor: AudioCompressor,
	AudioDeviceInput: AudioDeviceInput,
	AudioDeviceOutput: AudioDeviceOutput,
	AudioDistortion: AudioDistortion,
	AudioEcho: AudioEcho,
	AudioEmitter: AudioEmitter,
	AudioEqualizer: AudioEqualizer,
	AudioFader: AudioFader,
	AudioFlanger: AudioFlanger,
	AudioListener: AudioListener,
	AudioPitchShifter: AudioPitchShifter,
	AudioPlayer: AudioPlayer,
	AudioReverb: AudioReverb,
	AudioSearchParams: AudioSearchParams,
	AvatarChatService: AvatarChatService,
	AvatarEditorService: AvatarEditorService,
	AvatarImportService: AvatarImportService,
	Backpack: Backpack,
	BadgeService: BadgeService,
	BaseImportData: BaseImportData,
	AnimationImportData: AnimationImportData,
	FacsImportData: FacsImportData,
	GroupImportData: GroupImportData,
	JointImportData: JointImportData,
	MaterialImportData: MaterialImportData,
	MeshImportData: MeshImportData,
	RootImportData: RootImportData,
	BasePlayerGui: BasePlayerGui,
	CoreGui: CoreGui,
	PlayerGui: PlayerGui,
	StarterGui: StarterGui,
	BaseWrap: BaseWrap,
	WrapLayer: WrapLayer,
	WrapTarget: WrapTarget,
	Beam: Beam,
	BindableEvent: BindableEvent,
	BindableFunction: BindableFunction,
	BodyMover: BodyMover,
	BodyAngularVelocity: BodyAngularVelocity,
	BodyForce: BodyForce,
	BodyGyro: BodyGyro,
	BodyPosition: BodyPosition,
	BodyThrust: BodyThrust,
	BodyVelocity: BodyVelocity,
	RocketPropulsion: RocketPropulsion,
	Breakpoint: Breakpoint,
	BrowserService: BrowserService,
	BubbleChatMessageProperties: BubbleChatMessageProperties,
	BulkImportService: BulkImportService,
	CacheableContentProvider: CacheableContentProvider,
	HSRDataContentProvider: HSRDataContentProvider,
	MeshContentProvider: MeshContentProvider,
	SolidModelContentProvider: SolidModelContentProvider,
	CalloutService: CalloutService,
	Camera: Camera,
	CaptureService: CaptureService,
	ChangeHistoryService: ChangeHistoryService,
	CharacterAppearance: CharacterAppearance,
	BodyColors: BodyColors,
	CharacterMesh: CharacterMesh,
	Clothing: Clothing,
	Pants: Pants,
	Shirt: Shirt,
	ShirtGraphic: ShirtGraphic,
	Skin: Skin,
	Chat: Chat,
	ClickDetector: ClickDetector,
	DragDetector: DragDetector,
	Clouds: Clouds,
	ClusterPacketCache: ClusterPacketCache,
	CollectionService: CollectionService,
	Configuration: Configuration,
	ConfigureServerService: ConfigureServerService,
	Constraint: Constraint,
	AlignOrientation: AlignOrientation,
	AlignPosition: AlignPosition,
	AngularVelocity: AngularVelocity,
	AnimationConstraint: AnimationConstraint,
	BallSocketConstraint: BallSocketConstraint,
	HingeConstraint: HingeConstraint,
	LineForce: LineForce,
	LinearVelocity: LinearVelocity,
	PlaneConstraint: PlaneConstraint,
	Plane: Plane,
	RigidConstraint: RigidConstraint,
	RodConstraint: RodConstraint,
	RopeConstraint: RopeConstraint,
	SlidingBallConstraint: SlidingBallConstraint,
	CylindricalConstraint: CylindricalConstraint,
	PrismaticConstraint: PrismaticConstraint,
	SpringConstraint: SpringConstraint,
	Torque: Torque,
	TorsionSpringConstraint: TorsionSpringConstraint,
	UniversalConstraint: UniversalConstraint,
	VectorForce: VectorForce,
	ContentProvider: ContentProvider,
	ContextActionService: ContextActionService,
	Controller: Controller,
	HumanoidController: HumanoidController,
	SkateboardController: SkateboardController,
	VehicleController: VehicleController,
	ControllerBase: ControllerBase,
	AirController: AirController,
	ClimbController: ClimbController,
	GroundController: GroundController,
	SwimController: SwimController,
	ControllerManager: ControllerManager,
	ControllerService: ControllerService,
	CookiesService: CookiesService,
	CorePackages: CorePackages,
	CoreScriptDebuggingManagerHelper: CoreScriptDebuggingManagerHelper,
	CoreScriptSyncService: CoreScriptSyncService,
	CrossDMScriptChangeListener: CrossDMScriptChangeListener,
	CustomEvent: CustomEvent,
	CustomEventReceiver: CustomEventReceiver,
	DataModelMesh: DataModelMesh,
	BevelMesh: BevelMesh,
	BlockMesh: BlockMesh,
	CylinderMesh: CylinderMesh,
	FileMesh: FileMesh,
	SpecialMesh: SpecialMesh,
	DataModelPatchService: DataModelPatchService,
	DataModelSession: DataModelSession,
	DataStoreIncrementOptions: DataStoreIncrementOptions,
	DataStoreInfo: DataStoreInfo,
	DataStoreKey: DataStoreKey,
	DataStoreKeyInfo: DataStoreKeyInfo,
	DataStoreObjectVersionInfo: DataStoreObjectVersionInfo,
	DataStoreOptions: DataStoreOptions,
	DataStoreService: DataStoreService,
	DataStoreSetOptions: DataStoreSetOptions,
	Debris: Debris,
	DebugSettings: DebugSettings,
	DebuggablePluginWatcher: DebuggablePluginWatcher,
	DebuggerBreakpoint: DebuggerBreakpoint,
	DebuggerConnection: DebuggerConnection,
	LocalDebuggerConnection: LocalDebuggerConnection,
	DebuggerConnectionManager: DebuggerConnectionManager,
	DebuggerLuaResponse: DebuggerLuaResponse,
	DebuggerManager: DebuggerManager,
	DebuggerUIService: DebuggerUIService,
	DebuggerVariable: DebuggerVariable,
	DebuggerWatch: DebuggerWatch,
	DeviceIdService: DeviceIdService,
	Dialog: Dialog,
	DialogChoice: DialogChoice,
	DraftsService: DraftsService,
	Dragger: Dragger,
	DraggerService: DraggerService,
	EulerRotationCurve: EulerRotationCurve,
	EventIngestService: EventIngestService,
	ExperienceAuthService: ExperienceAuthService,
	ExperienceInviteOptions: ExperienceInviteOptions,
	Explosion: Explosion,
	FaceAnimatorService: FaceAnimatorService,
	FaceControls: FaceControls,
	FaceInstance: FaceInstance,
	Decal: Decal,
	Texture: Texture,
	FacialAnimationRecordingService: FacialAnimationRecordingService,
	FacialAnimationStreamingServiceStats: FacialAnimationStreamingServiceStats,
	FacialAnimationStreamingServiceV2: FacialAnimationStreamingServiceV2,
	FacialAnimationStreamingSubsessionStats: FacialAnimationStreamingSubsessionStats,
	Feature: Feature,
	Hole: Hole,
	MotorFeature: MotorFeature,
	File: File,
	Fire: Fire,
	FlagStandService: FlagStandService,
	FloatCurve: FloatCurve,
	FlyweightService: FlyweightService,
	CSGDictionaryService: CSGDictionaryService,
	NonReplicatedCSGDictionaryService: NonReplicatedCSGDictionaryService,
	Folder: Folder,
	ForceField: ForceField,
	FriendService: FriendService,
	FunctionalTest: FunctionalTest,
	GamePassService: GamePassService,
	GameSettings: GameSettings,
	GamepadService: GamepadService,
	Geometry: Geometry,
	GeometryService: GeometryService,
	GetTextBoundsParams: GetTextBoundsParams,
	GlobalDataStore: GlobalDataStore,
	DataStore: DataStore,
	OrderedDataStore: OrderedDataStore,
	GroupService: GroupService,
	GuiBase: GuiBase,
	GuiBase2d: GuiBase2d,
	GuiObject: GuiObject,
	CanvasGroup: CanvasGroup,
	Frame: Frame,
	GuiButton: GuiButton,
	ImageButton: ImageButton,
	TextButton: TextButton,
	GuiLabel: GuiLabel,
	ImageLabel: ImageLabel,
	TextLabel: TextLabel,
	ScrollingFrame: ScrollingFrame,
	TextBox: TextBox,
	VideoFrame: VideoFrame,
	ViewportFrame: ViewportFrame,
	LayerCollector: LayerCollector,
	BillboardGui: BillboardGui,
	PluginGui: PluginGui,
	DockWidgetPluginGui: DockWidgetPluginGui,
	QWidgetPluginGui: QWidgetPluginGui,
	ScreenGui: ScreenGui,
	GuiMain: GuiMain,
	SurfaceGuiBase: SurfaceGuiBase,
	AdGui: AdGui,
	SurfaceGui: SurfaceGui,
	GuiBase3d: GuiBase3d,
	FloorWire: FloorWire,
	InstanceAdornment: InstanceAdornment,
	SelectionBox: SelectionBox,
	PVAdornment: PVAdornment,
	HandleAdornment: HandleAdornment,
	BoxHandleAdornment: BoxHandleAdornment,
	ConeHandleAdornment: ConeHandleAdornment,
	CylinderHandleAdornment: CylinderHandleAdornment,
	ImageHandleAdornment: ImageHandleAdornment,
	LineHandleAdornment: LineHandleAdornment,
	SphereHandleAdornment: SphereHandleAdornment,
	WireframeHandleAdornment: WireframeHandleAdornment,
	ParabolaAdornment: ParabolaAdornment,
	SelectionSphere: SelectionSphere,
	PartAdornment: PartAdornment,
	HandlesBase: HandlesBase,
	ArcHandles: ArcHandles,
	Handles: Handles,
	SurfaceSelection: SurfaceSelection,
	SelectionLasso: SelectionLasso,
	SelectionPartLasso: SelectionPartLasso,
	SelectionPointLasso: SelectionPointLasso,
	GuiService: GuiService,
	GuidRegistryService: GuidRegistryService,
	HapticService: HapticService,
	HeightmapImporterService: HeightmapImporterService,
	HiddenSurfaceRemovalAsset: HiddenSurfaceRemovalAsset,
	Highlight: Highlight,
	Hopper: Hopper,
	HttpRbxApiService: HttpRbxApiService,
	HttpRequest: HttpRequest,
	HttpService: HttpService,
	Humanoid: Humanoid,
	HumanoidDescription: HumanoidDescription,
	IKControl: IKControl,
	ILegacyStudioBridge: ILegacyStudioBridge,
	LegacyStudioBridge: LegacyStudioBridge,
	IXPService: IXPService,
	IncrementalPatchBuilder: IncrementalPatchBuilder,
	InputObject: InputObject,
	InsertService: InsertService,
	JointInstance: JointInstance,
	DynamicRotate: DynamicRotate,
	RotateP: RotateP,
	RotateV: RotateV,
	Glue: Glue,
	ManualSurfaceJointInstance: ManualSurfaceJointInstance,
	ManualGlue: ManualGlue,
	ManualWeld: ManualWeld,
	Motor: Motor,
	Motor6D: Motor6D,
	Rotate: Rotate,
	Snap: Snap,
	VelocityMotor: VelocityMotor,
	Weld: Weld,
	JointsService: JointsService,
	KeyboardService: KeyboardService,
	Keyframe: Keyframe,
	KeyframeMarker: KeyframeMarker,
	KeyframeSequenceProvider: KeyframeSequenceProvider,
	LSPFileSyncService: LSPFileSyncService,
	LanguageService: LanguageService,
	Light: Light,
	PointLight: PointLight,
	SpotLight: SpotLight,
	SurfaceLight: SurfaceLight,
	Lighting: Lighting,
	LiveScriptingService: LiveScriptingService,
	LocalStorageService: LocalStorageService,
	AppStorageService: AppStorageService,
	UserStorageService: UserStorageService,
	LocalizationService: LocalizationService,
	LocalizationTable: LocalizationTable,
	CloudLocalizationTable: CloudLocalizationTable,
	LodDataEntity: LodDataEntity,
	LodDataService: LodDataService,
	LogService: LogService,
	LoginService: LoginService,
	LuaSettings: LuaSettings,
	LuaSourceContainer: LuaSourceContainer,
	BaseScript: BaseScript,
	CoreScript: CoreScript,
	Script: Script,
	LocalScript: LocalScript,
	ModuleScript: ModuleScript,
	LuaWebService: LuaWebService,
	LuauScriptAnalyzerService: LuauScriptAnalyzerService,
	MarkerCurve: MarkerCurve,
	MarketplaceService: MarketplaceService,
	MaterialGenerationService: MaterialGenerationService,
	MaterialService: MaterialService,
	MaterialVariant: MaterialVariant,
	MemStorageConnection: MemStorageConnection,
	MemStorageService: MemStorageService,
	MemoryStoreQueue: MemoryStoreQueue,
	MemoryStoreService: MemoryStoreService,
	MemoryStoreSortedMap: MemoryStoreSortedMap,
	Message: Message,
	Hint: Hint,
	MessageBusConnection: MessageBusConnection,
	MessageBusService: MessageBusService,
	MessagingService: MessagingService,
	MetaBreakpoint: MetaBreakpoint,
	MetaBreakpointContext: MetaBreakpointContext,
	MetaBreakpointManager: MetaBreakpointManager,
	Mouse: Mouse,
	PlayerMouse: PlayerMouse,
	PluginMouse: PluginMouse,
	MouseService: MouseService,
	MultipleDocumentInterfaceInstance: MultipleDocumentInterfaceInstance,
	NetworkMarker: NetworkMarker,
	NetworkPeer: NetworkPeer,
	NetworkClient: NetworkClient,
	NetworkServer: NetworkServer,
	NetworkReplicator: NetworkReplicator,
	ClientReplicator: ClientReplicator,
	ServerReplicator: ServerReplicator,
	NetworkSettings: NetworkSettings,
	NoCollisionConstraint: NoCollisionConstraint,
	NotificationService: NotificationService,
	OmniRecommendationsService: OmniRecommendationsService,
	OpenCloudApiV1: OpenCloudApiV1,
	OpenCloudService: OpenCloudService,
	PVInstance: PVInstance,
	BasePart: BasePart,
	CornerWedgePart: CornerWedgePart,
	FormFactorPart: FormFactorPart,
	Part: Part,
	FlagStand: FlagStand,
	Platform: Platform,
	Seat: Seat,
	SkateboardPlatform: SkateboardPlatform,
	SpawnLocation: SpawnLocation,
	WedgePart: WedgePart,
	Terrain: Terrain,
	TriangleMeshPart: TriangleMeshPart,
	MeshPart: MeshPart,
	PartOperation: PartOperation,
	IntersectOperation: IntersectOperation,
	NegateOperation: NegateOperation,
	UnionOperation: UnionOperation,
	TrussPart: TrussPart,
	VehicleSeat: VehicleSeat,
	Model: Model,
	Actor: Actor,
	BackpackItem: BackpackItem,
	HopperBin: HopperBin,
	Tool: Tool,
	Flag: Flag,
	Status: Status,
	WorldRoot: WorldRoot,
	Workspace: Workspace,
	WorldModel: WorldModel,
	PackageLink: PackageLink,
	PackageService: PackageService,
	PackageUIService: PackageUIService,
	Pages: Pages,
	AudioPages: AudioPages,
	CatalogPages: CatalogPages,
	DataStoreKeyPages: DataStoreKeyPages,
	DataStoreListingPages: DataStoreListingPages,
	DataStorePages: DataStorePages,
	DataStoreVersionPages: DataStoreVersionPages,
	FriendPages: FriendPages,
	InventoryPages: InventoryPages,
	EmotesPages: EmotesPages,
	OutfitPages: OutfitPages,
	StandardPages: StandardPages,
	PartOperationAsset: PartOperationAsset,
	ParticleEmitter: ParticleEmitter,
	PatchBundlerFileWatch: PatchBundlerFileWatch,
	PatchMapping: PatchMapping,
	Path: Path,
	PathfindingLink: PathfindingLink,
	PathfindingModifier: PathfindingModifier,
	PathfindingService: PathfindingService,
	PausedState: PausedState,
	PausedStateBreakpoint: PausedStateBreakpoint,
	PausedStateException: PausedStateException,
	PermissionsService: PermissionsService,
	PhysicsService: PhysicsService,
	PhysicsSettings: PhysicsSettings,
	PlaceStatsService: PlaceStatsService,
	PlacesService: PlacesService,
	Player: Player,
	PlayerEmulatorService: PlayerEmulatorService,
	PlayerScripts: PlayerScripts,
	Players: Players,
	Plugin: Plugin,
	PluginAction: PluginAction,
	PluginCapabilities: PluginCapabilities,
	PluginDebugService: PluginDebugService,
	PluginDragEvent: PluginDragEvent,
	PluginGuiService: PluginGuiService,
	PluginManagementService: PluginManagementService,
	PluginManager: PluginManager,
	PluginManagerInterface: PluginManagerInterface,
	PluginMenu: PluginMenu,
	PluginPolicyService: PluginPolicyService,
	PluginToolbar: PluginToolbar,
	PluginToolbarButton: PluginToolbarButton,
	PointsService: PointsService,
	PolicyService: PolicyService,
	PoseBase: PoseBase,
	NumberPose: NumberPose,
	Pose: Pose,
	PostEffect: PostEffect,
	BloomEffect: BloomEffect,
	BlurEffect: BlurEffect,
	ColorCorrectionEffect: ColorCorrectionEffect,
	DepthOfFieldEffect: DepthOfFieldEffect,
	SunRaysEffect: SunRaysEffect,
	ProcessInstancePhysicsService: ProcessInstancePhysicsService,
	ProximityPrompt: ProximityPrompt,
	ProximityPromptService: ProximityPromptService,
	PublishService: PublishService,
	RbxAnalyticsService: RbxAnalyticsService,
	ReflectionMetadata: ReflectionMetadata,
	ReflectionMetadataCallbacks: ReflectionMetadataCallbacks,
	ReflectionMetadataClasses: ReflectionMetadataClasses,
	ReflectionMetadataEnums: ReflectionMetadataEnums,
	ReflectionMetadataEvents: ReflectionMetadataEvents,
	ReflectionMetadataFunctions: ReflectionMetadataFunctions,
	ReflectionMetadataItem: ReflectionMetadataItem,
	ReflectionMetadataClass: ReflectionMetadataClass,
	ReflectionMetadataEnum: ReflectionMetadataEnum,
	ReflectionMetadataEnumItem: ReflectionMetadataEnumItem,
	ReflectionMetadataMember: ReflectionMetadataMember,
	ReflectionMetadataProperties: ReflectionMetadataProperties,
	ReflectionMetadataYieldFunctions: ReflectionMetadataYieldFunctions,
	RemoteCursorService: RemoteCursorService,
	RemoteDebuggerServer: RemoteDebuggerServer,
	RemoteEvent: RemoteEvent,
	RemoteFunction: RemoteFunction,
	RenderSettings: RenderSettings,
	RenderingTest: RenderingTest,
	ReplicatedFirst: ReplicatedFirst,
	ReplicatedStorage: ReplicatedStorage,
	RobloxPluginGuiService: RobloxPluginGuiService,
	RobloxReplicatedStorage: RobloxReplicatedStorage,
	RobloxServerStorage: RobloxServerStorage,
	RomarkService: RomarkService,
	RotationCurve: RotationCurve,
	RtMessagingService: RtMessagingService,
	RunService: RunService,
	RuntimeScriptService: RuntimeScriptService,
	SafetyService: SafetyService,
	ScreenshotHud: ScreenshotHud,
	ScriptBuilder: ScriptBuilder,
	SyncScriptBuilder: SyncScriptBuilder,
	ScriptChangeService: ScriptChangeService,
	ScriptCloneWatcher: ScriptCloneWatcher,
	ScriptCloneWatcherHelper: ScriptCloneWatcherHelper,
	ScriptCommitService: ScriptCommitService,
	ScriptContext: ScriptContext,
	ScriptDebugger: ScriptDebugger,
	ScriptDocument: ScriptDocument,
	ScriptEditorService: ScriptEditorService,
	ScriptRegistrationService: ScriptRegistrationService,
	ScriptRuntime: ScriptRuntime,
	ScriptService: ScriptService,
	Selection: Selection,
	SelectionHighlightManager: SelectionHighlightManager,
	SensorBase: SensorBase,
	BuoyancySensor: BuoyancySensor,
	ControllerSensor: ControllerSensor,
	ControllerPartSensor: ControllerPartSensor,
	ServerScriptService: ServerScriptService,
	ServerStorage: ServerStorage,
	ServiceProvider: ServiceProvider,
	DataModel: DataModel,
	GenericSettings: GenericSettings,
	GlobalSettings: GlobalSettings,
	UserSettings: UserSettings,
	ServiceVisibilityService: ServiceVisibilityService,
	SessionService: SessionService,
	SharedTableRegistry: SharedTableRegistry,
	Sky: Sky,
	Smoke: Smoke,
	SmoothVoxelsUpgraderService: SmoothVoxelsUpgraderService,
	SnippetService: SnippetService,
	SocialService: SocialService,
	Sound: Sound,
	SoundEffect: SoundEffect,
	ChorusSoundEffect: ChorusSoundEffect,
	CompressorSoundEffect: CompressorSoundEffect,
	CustomSoundEffect: CustomSoundEffect,
	AssetSoundEffect: AssetSoundEffect,
	ChannelSelectorSoundEffect: ChannelSelectorSoundEffect,
	DistortionSoundEffect: DistortionSoundEffect,
	EchoSoundEffect: EchoSoundEffect,
	EqualizerSoundEffect: EqualizerSoundEffect,
	FlangeSoundEffect: FlangeSoundEffect,
	PitchShiftSoundEffect: PitchShiftSoundEffect,
	ReverbSoundEffect: ReverbSoundEffect,
	TremoloSoundEffect: TremoloSoundEffect,
	SoundGroup: SoundGroup,
	SoundService: SoundService,
	Sparkles: Sparkles,
	SpawnerService: SpawnerService,
	StackFrame: StackFrame,
	StandalonePluginScripts: StandalonePluginScripts,
	StarterGear: StarterGear,
	StarterPack: StarterPack,
	StarterPlayer: StarterPlayer,
	StarterPlayerScripts: StarterPlayerScripts,
	StarterCharacterScripts: StarterCharacterScripts,
	Stats: Stats,
	StatsItem: StatsItem,
	RunningAverageItemDouble: RunningAverageItemDouble,
	RunningAverageItemInt: RunningAverageItemInt,
	RunningAverageTimeIntervalItem: RunningAverageTimeIntervalItem,
	TotalCountTimeIntervalItem: TotalCountTimeIntervalItem,
	StopWatchReporter: StopWatchReporter,
	Studio: Studio,
	StudioAssetService: StudioAssetService,
	StudioCallout: StudioCallout,
	StudioData: StudioData,
	StudioDeviceEmulatorService: StudioDeviceEmulatorService,
	StudioObjectBase: StudioObjectBase,
	StudioWidget: StudioWidget,
	StudioPublishService: StudioPublishService,
	StudioScriptDebugEventListener: StudioScriptDebugEventListener,
	StudioSdkService: StudioSdkService,
	StudioService: StudioService,
	StudioTheme: StudioTheme,
	StyleBase: StyleBase,
	StyleRule: StyleRule,
	StyleSheet: StyleSheet,
	StyleDerive: StyleDerive,
	StyleLink: StyleLink,
	StylingService: StylingService,
	SurfaceAppearance: SurfaceAppearance,
	TaskScheduler: TaskScheduler,
	Team: Team,
	TeamCreateData: TeamCreateData,
	TeamCreatePublishService: TeamCreatePublishService,
	TeamCreateService: TeamCreateService,
	Teams: Teams,
	TeleportAsyncResult: TeleportAsyncResult,
	TeleportOptions: TeleportOptions,
	TeleportService: TeleportService,
	TemporaryCageMeshProvider: TemporaryCageMeshProvider,
	TemporaryScriptService: TemporaryScriptService,
	TerrainDetail: TerrainDetail,
	TerrainRegion: TerrainRegion,
	TestService: TestService,
	TextBoxService: TextBoxService,
	TextChannel: TextChannel,
	TextChatCommand: TextChatCommand,
	TextChatConfigurations: TextChatConfigurations,
	BubbleChatConfiguration: BubbleChatConfiguration,
	ChatInputBarConfiguration: ChatInputBarConfiguration,
	ChatWindowConfiguration: ChatWindowConfiguration,
	TextChatMessage: TextChatMessage,
	TextChatMessageProperties: TextChatMessageProperties,
	TextChatService: TextChatService,
	TextFilterResult: TextFilterResult,
	TextFilterTranslatedResult: TextFilterTranslatedResult,
	TextService: TextService,
	TextSource: TextSource,
	ThirdPartyUserService: ThirdPartyUserService,
	ThreadState: ThreadState,
	TimerService: TimerService,
	ToastNotificationService: ToastNotificationService,
	TouchInputService: TouchInputService,
	TouchTransmitter: TouchTransmitter,
	TracerService: TracerService,
	TrackerLodController: TrackerLodController,
	TrackerStreamAnimation: TrackerStreamAnimation,
	Trail: Trail,
	Translator: Translator,
	TutorialService: TutorialService,
	TweenBase: TweenBase,
	Tween: Tween,
	TweenService: TweenService,
	UGCAvatarService: UGCAvatarService,
	UGCValidationService: UGCValidationService,
	UIBase: UIBase,
	UIComponent: UIComponent,
	UIConstraint: UIConstraint,
	UIAspectRatioConstraint: UIAspectRatioConstraint,
	UISizeConstraint: UISizeConstraint,
	UITextSizeConstraint: UITextSizeConstraint,
	UICorner: UICorner,
	UIGradient: UIGradient,
	UILayout: UILayout,
	UIGridStyleLayout: UIGridStyleLayout,
	UIGridLayout: UIGridLayout,
	UIListLayout: UIListLayout,
	UIPageLayout: UIPageLayout,
	UITableLayout: UITableLayout,
	UIPadding: UIPadding,
	UIScale: UIScale,
	UIStroke: UIStroke,
	UnvalidatedAssetService: UnvalidatedAssetService,
	UserGameSettings: UserGameSettings,
	UserInputService: UserInputService,
	UserService: UserService,
	VRService: VRService,
	VRStatusService: VRStatusService,
	ValueBase: ValueBase,
	BinaryStringValue: BinaryStringValue,
	BoolValue: BoolValue,
	BrickColorValue: BrickColorValue,
	CFrameValue: CFrameValue,
	Color3Value: Color3Value,
	DoubleConstrainedValue: DoubleConstrainedValue,
	IntConstrainedValue: IntConstrainedValue,
	IntValue: IntValue,
	NumberValue: NumberValue,
	ObjectValue: ObjectValue,
	RayValue: RayValue,
	StringValue: StringValue,
	Vector3Value: Vector3Value,
	Vector3Curve: Vector3Curve,
	VersionControlService: VersionControlService,
	VideoCaptureService: VideoCaptureService,
	VideoService: VideoService,
	VirtualInputManager: VirtualInputManager,
	VirtualUser: VirtualUser,
	VisibilityCheckDispatcher: VisibilityCheckDispatcher,
	Visit: Visit,
	VoiceChatInternal: VoiceChatInternal,
	VoiceChatService: VoiceChatService,
	WeldConstraint: WeldConstraint,
	Wire: Wire,
}

--yummers thanks ffrostfall might revise later for my usecases
type UnknownConnectCallback = (...any) -> ()
type Signal = RBXScriptSignal

type function get_connect(v: type): type?
	local connect = v:readproperty(types.singleton("Connect"))

	if connect and connect:is("function") then
		local head = connect:parameters().head

		if not head then
			return nil
		end
		local callback = head[2]

		if callback and callback:is("function") then
			return callback
		end
	end

	return nil
end

type function is_signal(v: type): type?
	if v == Signal and v.tag == "class" then
		return UnknownConnectCallback
	-- luau lsp types RBXScriptSignals as tables
	elseif v.tag == "table" then
		local connect = get_connect(v)

		if connect then
			return connect
		end

		-- i cba to test if this is needed or not
		local metatable = v:metatable()

		if metatable then
			connect = get_connect(metatable)

			if connect then
				return connect
			end
		end
	end

	return nil
end

type instance = Instance

local action_metatable_symbol = table.freeze({})
type Action = setmetatable<{
	callback: (inst: Instance) -> (),
	priority: number,
}, typeof(action_metatable_symbol)>

type function GetInstance(class_name: type, instances: type): type
	if not class_name:is("singleton") then
		return instance
	end

	return instances:readproperty(class_name) or instance
end

export type function PropertiesFor(class_name: type): type
	local object = GetInstance(class_name, Instances)
	local head_table = {}
	local properties = types.newtable()
	local current = object

	local read_only_properties = table.freeze({
		types.singleton("RobloxLocked"),
		types.singleton("ClassName"),
		types.singleton("UniqueId"),
	})

	while current do
		for property, value in current:properties() do
			if table.find(read_only_properties, property) then
				continue
			end

			local value_read = value.read :: type
			local callback = is_signal(value_read)

			if callback then
				--properties:setproperty(property :: any, types.optional(callback))
				continue
			elseif value_read.tag == "function" or value_read.tag == "intersection" then
				continue
			elseif value.write then
				head_table[1] = value_read
				properties:setproperty(property :: any, types.optional(value.write :: any))
			end
		end

		current = current:readparent()
	end

	local nested_table_or_function_or_child = types.newtable()
	local function_returns_child = types.newfunction()
	local union = types.unionof(nested_table_or_function_or_child, function_returns_child, instance, Action)

	nested_table_or_function_or_child:setindexer(types.number, union)
	function_returns_child:setreturns({ union })

	properties:setindexer(types.number, types.optional(union))

	return properties
end

local function create<ClassName>(
	className: ClassName | keyof<Instances>
): (props: PropertiesFor<ClassName>) -> GetInstance<ClassName, Instances>
	return function(properties: {[string]: any})
		local success, callResult = pcall(Instance.new, className)
		assert(success, callResult)
		for property, value in properties do
			callResult[property] = value
		end
		return callResult
	end
end

return create
